### DISCUSSION TITLE
Using different kyber implementations with strongswan

### DISCUSSION THREAD

**Discussion by yamus-didine** [open-quantum-safe/liboqs#1520] [Q&A] [answered]
Hello.
I am using Kyber KEM with Strongswan version 6.0.0beta3 and Liboqs 0.7.2
My inquiry is that after compiling Liboqs with strongswan, AVX2 implementation is the default.
I want to be able to use all Kyber implementations (Reference, 90s and AVX2 90s) alongside the default AVX2 implementation.
Is there a way to do it?

**Comment by baentsch** [comment]
> Hello. I am using Kyber KEM with Strongswan version 6.0.0beta3 and Liboqs 0.7.2 

Thanks for letting us know about your use case. Please consider contributing your integration to [oqs-demos](https://github.com/open-quantum-safe/oqs-demos).

> My inquiry is that after compiling Liboqs with strongswan, AVX2 implementation is the default. I want to be able to use all Kyber implementations (Reference, 90s and AVX2 90s) alongside the default AVX2 implementation. Is there a way to do it?

Yes, (for `x64`) you can build the code without AVX2 support (i.e., the reference code) by setting [OQS_OPT_TARGET](https://github.com/open-quantum-safe/liboqs/blob/main/CONFIGURE.md#oqs_opt_target) to the value `generic` to disable any CPU optimization (code).

**Reply by dstebila** [reply]
We do not, however, support the 90s versions anymore, as they have not been selected for standardization by NIST.

**Reply by yamus-didine** [reply]
> We do not, however, support the 90s versions anymore, as they have not been selected for standardization by NIST.

Yes. I am aware of that. The thing is that I am curious to compare different Kyber implementations.
Thank you for your reply professor.

**Comment by yamus-didine** [comment]
> Yes, (for `x64`) you can build the code without AVX2 support (i.e., the reference code) by setting [OQS_OPT_TARGET](https://github.com/open-quantum-safe/liboqs/blob/main/CONFIGURE.md#oqs_opt_target) to the value `generic` to disable any CPU optimization (code).

So, I did what you suggested. However, It seems it did not work for me, and I don't know why.
Here is how I built Liboqs.

- In the Liboqs folder, I created a separate `build/` directory.
- In `build/`, I issued the following
`cmake -GNinja -DOQS_USE_OPENSSL=ON -DBUILD_SHARED_LIBS=ON -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=Release -DOQS_DIST_BUILD=OFF -DOQS_OPT_TARGET=generic -DOQS_BUILD_ONLY_LIB=ON .. && ninja && ninja install`

To check my work, I made `crypto_kem_keypair(...)` function (in the `kem.c` file for each Kyber implementation) print a string (to know precisely which Kyber implementation is being used. However, I get the same thing. Always, Avx2 implementation is being called. Is there a problem somewhere?

Should I edit the file `CMakeCache.txt` and then re-build Liboqs as I did in the following?

![kyber_versions](https://github.com/open-quantum-safe/liboqs/assets/134303261/4f758330-29fa-451a-97a0-4995931b9fce)

**Reply by baentsch** [reply, marked as answer]
> Should I edit the file CMakeCache.txt and then re-build Liboqs as I did in the following?

Before issuing any `cmake` command, you should delete the "build" directory to be sure no `cmake` artifacts (incl. caches) remain.
